File: src/PlatynUI/technology/uiautomation/core/keyboarddevice.py
# SPDX-FileCopyrightText: 2024 Daniel Biehl <daniel.biehl@imbus.de>
#
# SPDX-License-Identifier: Apache-2.0

from typing import Optional

from PlatynUI.ui.core.devices.basekeyboarddevice import BaseKeyboardDevice, BaseKeyCode, InputType, Key

from .loader import DotNetInterface


class UiaKeyboardDevice(BaseKeyboardDevice):
    def __init__(self) -> None:
        super().__init__()
        self.input_type: Optional[InputType] = None

    def key_to_keycode(self, key: Key) -> BaseKeyCode:
        result = DotNetInterface().keyboard_device().KeyToKeyCode(key)
        return BaseKeyCode(result.Key, result.Code, result.Valid, result.ErrorText)

    def send_keycode(self, keycode: BaseKeyCode, pressed: bool) -> bool:
        if not keycode.valid:
            raise ValueError(f"Invalid key: {keycode}")

        return DotNetInterface().keyboard_device().SendKeyCode(keycode.code, pressed)

File: src/PlatynUI.Runtime/KeyboardDevice.cs
// SPDX-FileCopyrightText: 2024 Daniel Biehl <daniel.biehl@imbus.de>
//
// SPDX-License-Identifier: Apache-2.0

using System.ComponentModel.Composition;
using PlatynUI.Runtime.Core;

namespace PlatynUI.Runtime;

public class KeyboardDevice
{
    [Import]
    protected IKeyboardDevice? keyboardDevice;

    private KeyboardDevice()
    {
        PlatynUiExtensions.ComposeParts(this);
    }

    private static KeyboardDevice? _instance;
    private static KeyboardDevice Instance => _instance ??= new KeyboardDevice();

    public static Keycode KeyToKeyCode(object? key)
    {
        return Instance.keyboardDevice?.KeyToKeyCode(key) ?? new Keycode(null, null, false, "No keyboard device found");
    }

    public static bool SendKeyCode(object keyCode, bool pressed)
    {
        return Instance.keyboardDevice?.SendKeyCode(keyCode, pressed) ?? false;
    }
}
